File: 3020-find-the-maximum-number-of-elements-in-subset/3020-find-the-maximum-number-of-elements-in-subset.java
class Solution {
    public int maximumLength(int[] nums) {
        HashMap<Integer, Integer> map = new HashMap<>();
        int res = 1;
        boolean flag = false;
        
        for (int num : nums)
            map.put(num, map.getOrDefault(num, 0) + 1);
        
        for (int key : map.keySet()) {
            if (key == 1) {
                flag = true;
                continue;
            }
            int num = key;
            int cnt = 0;
            int i = 1;
            while (map.getOrDefault((int)Math.pow(num, i), 0) >= 2) {
                i *= 2;
                cnt += 2;
            }
            if (map.getOrDefault((int)Math.pow(num, i), 0) == 1) {
                cnt += 1;
            } else {
                cnt -= 1;
            }
            res = Math.max(res, cnt);
        }
        
        if (flag) {
            int cnt = map.get(1);
            if (cnt % 2 == 0) cnt --;
            res = Math.max(res, cnt);
        }
        return res;        
    }
}

File: 3020-find-the-maximum-number-of-elements-in-subset/3020-find-the-maximum-number-of-elements-in-subset.py
class Solution:
    def maximumLength(self, nums: List[int]) -> int:
        m = collections.defaultdict(int)
        for i in nums:
            m[i] += 1
        ans = m[1]
        if m[1]>0 and m[1]%2 == 0:
            ans = m[1]-1
        for i in sorted(nums):
            res = 1
            if i==1:
                continue
            while m[i**2]>0 and m[i]>1:
                res+=2
                m[i]-=2
                i = i**2
            ans = max(ans, res)
        return ans
        